xwiki-pro-macros-api/src/main/java/com/xwiki/macros/internal/updateReferences/ReferencesSolrMetadataExtractor.java

```diff
@@ -0,0 +1,202 @@
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.macros.internal.updateReferences;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.solr.common.SolrInputDocument;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.DocumentReferenceResolver;
+import org.xwiki.model.validation.EntityNameValidationConfiguration;
+import org.xwiki.model.validation.EntityNameValidationManager;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.XDOM;
+import org.xwiki.rendering.block.match.MacroBlockMatcher;
+import org.xwiki.search.solr.SolrEntityMetadataExtractor;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * Handles the backlinks of the reference parameters of macros. Ensure that the parameters that are references always
+ * have a valid reference even if the page is renamed or moved.
+ *
+ * @version $Id$
+ * @since 1.29
+ */
+@Component
+@Named("macrosReferencesUpdate")
+@Singleton
+public class ReferencesSolrMetadataExtractor implements SolrEntityMetadataExtractor<XWikiDocument>
+{
+    private static final Map<String, List<String>> MACROS_TO_SEARCH = Map.of(
+        "button", List.of("url"),
+        "confluence_pagetree", List.of("root")
+    );
+
+    @Inject
+    @Named("explicit")
+    private DocumentReferenceResolver<String> explicitDocumentReferenceResolver;
+
+    @Inject
+    private LinkRegistry linkRegistry;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private EntityNameValidationManager entityNameValidationManager;
+
+    @Inject
+    private EntityNameValidationConfiguration entityNameValidationConfiguration;
+
+    @Inject
+    private Logger logger;
+
+    @Override
+    public boolean extract(XWikiDocument document, SolrInputDocument solrDocument)
+    {
+        XDOM xdom = document.getXDOM();
+        boolean linksRegistered = false;
+
+        for (Map.Entry<String, List<String>> macroToSearch : MACROS_TO_SEARCH.entrySet()) {
```

Comment:
I think this misses macros that are in other macros. If I'm not wrong, you have to recursively parse the content of all the macros and add your specific handling for the macros you want to handle. 

See for instance https://snippets.xwiki.org/xwiki/bin/view/Extension/RemoveMacrosFromXDOM/ for an example of code that searches a document content for macros.

Reply:
> I think using `Block.Axes.DESCENDANT` as a parameter solves the problem!

No, it doesn't. Implementing `MacroRefactoring` would solve this as then this whole complexity is handled by XWiki and you don't need to re-implement it.